Modernization

IvanKiral opened the pull request:
### Motivation

Which issue does this fix? Fixes #`issue number`

If no issue exists, what is the fix or new feature? Were there any reasons to fix/implement things that are not obvious?

### Checklist

- [ ] Code follows coding conventions held in this repo
- [ ] Automated tests have been added
- [ ] Tests are passing
- [ ] Docs have been updated (if applicable)
- [ ] Temporary settings (e.g. variables used during development and testing) have been reverted to defaults

### How to test

If manual testing is required, what are the steps?

IvanKiral committed:
remove support for legacy webhooks

IvanKiral committed:
change to biome, eslint and vitest

IvanKiral committed:
get rid of barrel files

IvanKiral committed:
rewrite from oop class to functional approach

IvanKiral committed:
add type guards

IvanKiral committed:
update public API

IvanKiral committed:
remove type guards

IvanKiral committed:
update notifications to have object_type at top level

IvanKiral committed:
update validation with better error handling

IvanKiral committed:
update signature helper

IvanKiral committed:
update tests

IvanKiral committed:
update workflows

IvanKiral committed:
move test to top-level

IvanKiral committed:
remove unnecesary libs and scripts from package.json

IvanKiral committed:
update readme

IvanKiral committed:
update from SafeParseResult to ParseResult

IvanKiral committed:
4.0.0-0

IvanKiral committed:
fix wrong import

IvanKiral committed:
add release workflow

IvanKiral committed:
audit deps

JiriLojda commented on lib/models/parse-webhook.ts:
```diff
@@ -0,0 +1,14 @@
+import { fromZodError } from "zod-validation-error";
+import { type WebhookResponse, webhookResponseSchema } from "./webhook-schemas.js";
+
+export type ParseResult<T> = { success: true; data: T } | { success: false; error: Error };
```
Would move this into a separate utility-type file as it is reused in other files.

JiriLojda reviewed (approved):
Consider either removing the changelog file or mentioning that the changelog was moved to releases.

IvanKiral committed:
move ParseResult type to utility file

Moves the shared ParseResult<T> type to lib/utils/types.ts as it's
used by both parse-webhook and signature-helper modules.

IvanKiral committed:
update changelog to point to GitHub releases

Replaces changelog content with redirect note as the release history
is now tracked in GitHub Releases.

IvanKiral committed:
update to pnpm

Enngage commented on lib/utils/types.ts:
```diff
@@ -0,0 +1 @@
+export type ParseResult<T> = { success: true; data: T } | { success: false; error: Error };
```
Result props could be `readonly`, you might consider using custom error class instead of generic `Error`

Enngage commented on lib/signatures/signature-helper.ts:
```diff
@@ -0,0 +1,53 @@
+    return { success: false, error: new Error("Webhook signature validation failed") };
+  }
+
+  try {
```
It's a small repo so it doesn't really matter, but I like to use `tryCatch` utils to avoid using nesting try/catch blocks. Something like -> https://github.com/kontent-ai/core-sdk-js/blob/master/lib/utils/try.utils.ts#L24C17-L24C25

IvanKiral replied:
those utils are good, but I tihnk they are overkill for now. They might be introduced later if needed.

Enngage commented on lib/signatures/signature-helper.ts:
```diff
@@ -0,0 +1,53 @@
+  } catch (error) {
+    return {
+      success: false,
+      error: new Error(
```
I would create a custom Error class to distinguish from general `Error`

Enngage commented on lib/signatures/signature-helper.ts:
```diff
@@ -0,0 +1,53 @@
+  }
+};
+
+export const isSignatureValid = (
```
Might be personal preference, but I like to use destructured object params instead of (string, string, string) as it's easier to read and understand at a glance. Passing 3 strings next to each other might obscur their meaning unless everything is named nicely:)

Enngage commented on lib/signatures/signature-helper.ts:
```diff
@@ -0,0 +1,53 @@
+/**
+ * Normalizes line breaks in a string by replacing all line break variations with CRLF (`\r\n`).
+ */
+export const replaceLinebreaks = (data: string): string => data.replace(/[\r\n]+/gm, "\r\n");
```
Would it be possible to hide `replaceLinebreaks` so that clients wouldn't need to call it explicitely? So that you have `isSignatureValid(body, secret, signature)` instead of `isSignatureValid(replaceLinebreaks(body), secret, signature)`?

Then you might not need to export it in Public API at all.

Enngage commented on test/notifications-types.spec.ts:
```diff
@@ -0,0 +1,74 @@
+    const mockNotifications: WebhookNotification[] = [];
+
+    mockNotifications.forEach((notification) => {
+      switch (notification.object_type) {
```
No `ts-pattern`? :D I know, it's overkill, but still :D

IvanKiral replied:
It was the reason fo the transformation of object_type to top :D so you can easily use switch and don;'t have to introduce ts-pattern (though I love it, not every time you need it) :D

Enngage commented on lib/models/webhook-schemas.ts:
```diff
@@ -0,0 +1,228 @@
+import { z } from "zod";
+
+export const webhookDeliverySlotSchema = z.enum(["published", "preview"]);
```
I would consider separating zod schemas with types in files and would just export all models (*) in public API instead of listing them one by one when you know you want to export all of them anyways. At least thats what I did in toolkit :) https://github.com/kontent-ai/migration-toolkit/blob/7a5951422c87b6c9dcf74bc2e0bf4450260c85d4/lib/index.ts#L5

IvanKiral replied:
I acutally like havving written all of the types, so it is easier to know what public interface is. Do not know what solution is better ;D

IvanKiral committed:
use internally replaceLineBreaks

IvanKiral committed:
make ParseResult readonly

IvanKiral committed:
use destructured object instead  function params

IvanKiral committed:
update readme

IvanKiral committed:
4.0.0-1

IvanKiral committed:
add files to package.json

IvanKiral committed:
4.0.0
